### PR TITLE
system test

### PR DESCRIPTION
**Pull request by aleeusgr**
- [x] #22 
- [x] #34 
- [x] #35 
- [ ] describe cmm
- [ ] add data source
- [x] add handler to stream a media
- [x] add a handler 
- [x] change the testing environment https://github.com/tokio-rs/axum/tree/main/examples
- [x] #29 
- [x] #18 is required
- [x] #5 
- [x] #3 

https://docs.rs/axum/latest/axum/index.html#routing
https://github.com/ulid/spec

### CODE REVIEW

File: src/main.rs [hunk 1]
```diff
@@ -104,6 +107,9 @@ async fn main() -> Result<()> {
             .layer(CatchPanicLayer::custom(runtime::catch_panic))
             // Mark headers as sensitive on both requests and responses.
             .layer(SetSensitiveHeadersLayer::new([header::AUTHORIZATION]))
+            // here is openAPI with UAT:
+            // TODO: link docs on merge
+            // Here is where I add new functionality, right?
             .merge(SwaggerUi::new("/swagger-ui").url("/api-doc/openapi.json", ApiDoc::openapi()));
```
Comment (aleeusgr) [inline src/main.rs:118]: [router.merge](https://docs.rs/axum/latest/axum/struct.Router.html#method.merge)
Merge the paths and fallbacks of two routers into a single [Router](https://docs.rs/axum/latest/axum/struct.Router.html).

- [ ] ~~add a new [layer](https://docs.rs/axum/latest/axum/struct.Router.html#method.layer) to the router~~
- [x] add new route to the router
- [x] add route | layer to show hello-world at localhost:3000/hw
- [x] what is the difference bewteen merge, layer, route?
- [tower::Layer](https://docs.rs/tower-layer/0.3.2/x86_64-unknown-linux-gnu/tower_layer/trait.Layer.html)
- Merge two **routers** into one.
- Router.[route](https://docs.rs/axum/latest/axum/struct.Router.html#method.route)
- [ ] ~~handler responds with a picture~~


Router is the type for composing [handlers](https://docs.rs/axum/latest/axum/handler/index.html) and services.
In axum a `handler` is an async function that accepts zero or more [“extractors”](https://docs.rs/axum/latest/axum/extract/index.html) as arguments and returns something that can be converted [into a response](https://docs.rs/axum/latest/axum/response/index.html).

Handlers are where your application logic lives and axum applications are built by routing between handlers.

https://blog.logrocket.com/rust-axum-error-handling/
https://docs.rs/axum/latest/axum/index.html#routing

Comment (aleeusgr) [reply src/main.rs:113]: SwaggerUI is imported at L29;
ApiDoc at L31

Comment (aleeusgr) [reply src/main.rs:113]: ah, ok so if I add a new [layer](https://docs.rs/axum/latest/axum/struct.Router.html#method.layer) it should fit with everything else, right?

Comment (aleeusgr) [reply src/main.rs:113]: This can be used to add additional processing to a request for a group of routes.

**Review by aleeusgr — COMMENTED** [review]
- [x] add endpoint to current app
create a router ~~to `.layer` or `.merge` it with the app defined at L83;~~

File: src/main.rs (hunk 1 above)
Comment (aleeusgr) [reply src/main.rs:113]: https://docs.rs/axum/latest/axum/index.html#routing

Comment (aleeusgr) [reply src/main.rs:113]: In axum a “handler” is an async function that accepts zero or more [“extractors”](https://docs.rs/axum/latest/axum/extract/index.html) as arguments and returns something that can be converted [into a response](https://docs.rs/axum/latest/axum/response/index.html).

Handlers are where your application logic lives and axum applications are built by routing between handlers.

Comment (aleeusgr) [reply src/main.rs:113]: https://github.com/ulid/spec

Comment (aleeusgr) [reply src/main.rs:113]: https://blog.logrocket.com/rust-axum-error-handling/

File: src/main.rs [hunk 2]
```diff
@@ -78,16 +81,19 @@ async fn main() -> Result<()> {
...
             .route_layer(axum::middleware::from_fn(middleware::metrics::track))
+            // layer adds additional processing to a request for a group of routes
             .layer(Extension(env))
             // Include trace context as header into the response.
             .layer(response_with_trace_layer())
             // Opentelemetry tracing middleware.
             // This returns a `TraceLayer` configured to use
             // OpenTelemetry’s conventional span field names.
             .layer(opentelemetry_tracing_layer())
+            // TODO: search ulid
```
Comment (aleeusgr) [inline src/main.rs]: https://github.com/ulid/spec

**Review by aleeusgr — COMMENTED** [review]
https://blog.logrocket.com/rust-axum-error-handling/

**Review by aleeusgr — COMMENTED** [review]
- [x] test [the existing example](https://medium.com/@zacchaeusoluwole/build-a-real-time-webcam-streaming-server-in-rust-with-opencv-65b4d57cd22f)

```rust
use opencv::{
    core::{Mat, Vector}, imgcodecs, prelude::*, videoio,
};
```

```shell
$ cargo run
error: failed to run custom build command for `clang-sys v1.7.0`
```
[known error for opencv](https://stackoverflow.com/questions/76042739/rust-compilation-error-failed-to-run-custom-build-command-for-clang-sys-v1-6-1)

[source code](https://github.com/Zacchaeus-Oluwole/RUST-PROJECTS/blob/main/real-time_webcam_streaming_server/rcvweb/src/main.rs)

# alternative
- [ ] #22

**Comment by aleeusgr** [comment]
<details><summary>webcam management module</summary>
<p>

https://dev.to/dalalrohit/live-stream-your-webcam-to-html-page-3ehf

https://docs.rs/nokhwa/latest/nokhwa/

https://crates.io/keywords/webcam

https://www.reddit.com/r/rust/comments/3eab0m/how_do_i_capture_a_webcam_feed/

</p>
</details>

**Comment by aleeusgr** [comment]
- [ ] Quick example of axum + async-stream
https://github.com/fasterthanlime/stream-example/blob/main/src/main.rs
- [ ] axum server that serves a live stream of its own tracing log to an HTML page
https://github.com/thallada/axum-log-stream/blob/main/src/main.rs

File: src/main.rs [hunk 3]
```diff
@@ -104,15 +110,24 @@ async fn main() -> Result<()> {
             .layer(CatchPanicLayer::custom(runtime::catch_panic))
             // Mark headers as sensitive on both requests and responses.
             .layer(SetSensitiveHeadersLayer::new([header::AUTHORIZATION]))
+            .route("/say-hello", get(say_hello)) // change handler
```
Comment (aleeusgr) [inline src/main.rs]: - [x] add a route
- [x] change the route.
- [x] create handler, handler does X with stream. 

path is a string of path segments separated by /. Each segment can be either static, a capture, or a wildcard.

method_router is the [MethodRouter](https://docs.rs/axum/latest/axum/routing/method_routing/struct.MethodRouter.html) that should receive the request if the path matches path. method_router will commonly be a handler wrapped in a method router like [get](https://docs.rs/axum/latest/axum/routing/method_routing/fn.get.html). See [handler](https://docs.rs/axum/latest/axum/handler/index.html) for more details on handlers.

https://docs.rs/axum/latest/axum/struct.Router.html#method.route

Comment (aleeusgr) [reply src/main.rs]: `get(function)` should  work for a route:

```rust
    let app = Router::new()
        .route("/", get(serve_stream))
        .layer(Extension(tx));
    axum::Server::bind(&"[::]:8080".parse().unwrap())
        .serve(app.into_make_service())
        .await
        .unwrap();
}
``` 



https://github.com/fasterthanlime/stream-example/blob/main/src/main.rs

Comment (aleeusgr) [reply src/main.rs]: Here is a Router example to use:
```rust 
pub mod config;
pub mod handlers; // notice this line
pub mod log;
pub mod partials;
pub mod state;
. . . 
    let app = Router::new()
        .route("/", get(handlers::home::get))
        .route("/other", get(handlers::other::get))
        .route("/log", get(handlers::log::get))
        .route("/log/stream", get(handlers::log::stream)) // handlers is a "local module"
        .with_state(AppState {
            config,
            log_receiver,
        })
        .layer(ServiceBuilder::new().layer(TraceLayer::new_for_http()));
```
[handlers](https://github.com/thallada/axum-log-stream/tree/main/src/handlers)

https://github.com/thallada/axum-log-stream/blob/main/src/main.rs

Comment (aleeusgr) [reply src/main.rs]: <details><summary>Prompt: </summary>
<p>

in rust axum what a handler does to a stream?
Response: In Rust Axum, a handler is responsible for processing and handling incoming requests from clients in the form of streams (iterators or channels). A stream in this context refers to an asynchronous sequence of messages or events. When you create a handler function in Axum, it takes a `Handler` trait object as its input parameter which implements the `axum::handler_wrapper::HandlerWrapper`.

Here's what a typical Axum handler does with a stream:

1. Receive the incoming stream of messages or events from clients through an endpoint (either HTTP requests, WebSockets, or other protocols supported by Axum).
2. Process each message or event in the stream individually using the `axum::extract::FromRequest` trait to extract relevant information from the request and convert it into a structured data type like a `Request`.
3. Perform any necessary business logic, validation, or transformations on the extracted data. This may involve calling other functions, making database queries, or interacting with external services.
4. Respond to each message or event in the stream by generating an appropriate response (either synchronously or asynchronously). For example, if you're working with HTTP requests, this might include returning a JSON-encoded response or redirecting the client to another URL.
5. If necessary, handle any errors that occur during request processing and return an appropriate error message or status code to the client.
6. Continue processing messages in the stream until all have been handled or the stream is closed by the client.
7. Once all messages have been processed, close the connection with the client (if applicable) and return control back to Axum's runtime loop for handling future requests.

In summary, a handler in Rust Axum processes incoming streams of events or messages from clients, performs necessary operations on each message, and generates appropriate responses before moving on to the next message in the stream.


</p>
</details>

Comment (aleeusgr) [reply src/main.rs]: This is a handler that streams logs: 

```rust
pub async fn stream(
    State(log_receiver): State<Receiver<Bytes>>,
) -> Sse<impl Stream<Item = Result<Event, Infallible>>> {
    let log_stream = WatchStream::new(log_receiver);
    let log_stream = log_stream.map(|line| {
        Ok(Event::default().data(
            html! {
                turbo-stream action="append" target="log" {
                    template {
                        (PreEscaped(convert_escaped(from_utf8(&line).unwrap()).unwrap()))
                    }
                }
            }
            .into_string(),
        ))
    });
    Sse::new(log_stream).keep_alive(
        axum::response::sse::KeepAlive::new()
            .interval(Duration::from_secs(15))
            .text("keep-alive-text"),
    )
}
```

**Comment by aleeusgr** [comment]
```rust
[alex@nixos:~/workshop/efected-coto-emmory]$ cargo run
   Compiling efected-coto-emmory v0.1.0 (/home/alex/workshop/efected-coto-emmory)
error[E0583]: file not found for module `handlers`
  --> src/main.rs:45:1
   |
45 | pub mod handlers; // notice this line
   | ^^^^^^^^^^^^^^^^^
   |
   = help: to create the module `handlers`, create file "src/handlers.rs" or "src/handlers/mod.rs"

error[E0433]: failed to resolve: could not find `log` in `handlers`
   --> src/main.rs:115:48
    |
115 |             .route("/say-hello", get(handlers::log::say_hello)) // change handler
    |                                                ^^^ could not find `log` in `handlers`

Some errors have detailed explanations: E0433, E0583.
For more information about an error, try `rustc --explain E0433`.
error: could not compile `efected-coto-emmory` (bin "efected-coto-emmory-app") due to 2 previous errors
```

File: src/main.rs (hunk 3 above)
Comment (aleeusgr) [reply src/main.rs]:
```rust
async fn serve_stream(Extension(tx): Extension<broadcast::Sender<Msg>>) -> impl IntoResponse {
    let mut rx = tx.subscribe();


    let (s_tx, s_rx) = mpsc::channel(1);


    tokio::spawn(async move {
        loop {
            match rx.recv().await {
                Ok(msg) => match msg {
                    Msg::Tick => s_tx.send(Ok(Bytes::from_static(b"Tick\n"))).await.unwrap(),
                    Msg::Tock => s_tx.send(Ok(Bytes::from_static(b"Tock\n"))).await.unwrap(),
                },
                Err(err) => {
                    println!("body error: {err}");
                    return;
                }
            }
        }
    });


    let body = Body::wrap_stream::<_, _, std::io::Error>(ReceiverStream::new(s_rx));
    let mut res = Response::new(body);
    res.headers_mut()
        .insert("server", "super-sample-code".try_into().unwrap());
    res
}
```

**Comment by aleeusgr** [comment]
https://imfeld.dev/writing/process_streaming_uploads_with_axum

**Comment by aleeusgr** [comment]
write a handler, an async function that returns something that can be converted [into a response](https://docs.rs/axum/latest/axum/response/index.html).

Anything that implements [IntoResponse](https://docs.rs/axum/latest/axum/response/trait.IntoResponse.html) can be returned from a handler.
- [x] https://docs.rs/axum/latest/axum/response/index.html
- [x] https://stackoverflow.com/questions/75542752/send-a-dynamicimage-as-a-response-with-axum
- [ ] https://docs.rs/axum/latest/axum/extract/index.html

What about a stream, can stream implement IntoResponse? or how does it actually happen?

**Comment by aleeusgr** [comment]
https://doc.rust-lang.org/std/path/struct.PathBuf.html

**Comment by aleeusgr** [comment]
https://tokio.rs/tokio/tutorial/streams
